db/version_set.cc

```diff
@@ -571,14 +629,24 @@ class LevelIterator final : public InternalIterator {
   RangeDelAggregator* range_del_agg_;
   IteratorWrapper file_iter_;  // May be nullptr
   PinnedIteratorsManager* pinned_iters_mgr_;
+  RangeTombstone range_tombstone_;
 };
 
 void LevelIterator::Seek(const Slice& target) {
   size_t new_file_index = FindFile(icomparator_, *flevel_, target);
 
   InitFileIterator(new_file_index);
   if (file_iter_.iter() != nullptr) {
-    file_iter_.Seek(target);
+    // Before we seek the iterator, find the next non-deleted key.
+    InitRangeTombstone(ExtractUserKey(target));
+    if (range_tombstone_.seq_ > 0) {
+      file_iter_.Seek(tombstone_internal_end_key());
```

Comment:
This `Seek` may also be jumping one too far. I wonder if there is a way to rely on the `InitRangeTombstone` and `Seek` / `SeekForPrev` that happen in `SkipEmptyFile{Forward,Backward}`. This logic feels a bit redundant here and in other callers of `SkipEmptyFile{Forward,Backward}`.

Reply:
Yeah, I noticed the redundancy as well, but wasn't clever enough when I wrote this code to get rid of it. I'd like to decide on whether all of this logic belongs somewhere else (i.e. `BlockBasedTableIterator`) before polishing it too much.